accepted/2024/tensors/README.md

```diff
@@ -0,0 +1,424 @@
+
+Due to how generic inference works and there being no way to say that `<TTensor, T>` is really more `TTensor<T>`, we currently expose these as being over `ITensor<TSelf, T>`. This works fine for `Tensor<T>` since it is a class, but doesn't allow efficient reusue with `TensorSpan`. If C# had proper support for the "self constraint" or better generic inference support this might be a non-issue and we could expose it the more desirable way.
+
+`Tensor<T>` is notably explicitly sealed as extensibility must be an intentional design point. The extensibility point here is then `ITensor<TSelf, T>` and `Tensor<T>` is only our own friendly wrapper over arbitrary memory. This allows it to be more efficient and also more easily control lifetimes, ownership, and more freely extend the type in the future as needed.
+
+```csharp
+namespace System.Numerics.Tensors;
+
+public sealed class Tensor<T>
+    : IDisposable,
+      ITensor<Tensor<T>, T>
+    where T : IEquatable<T>, IEqualityOperators<T, T, bool>
+{
+    // IDisposable
+
+    void Dispose();
+
+    // ITensor
+
+    public static Tensor<T> Empty { get; }
+
+    public bool IsEmpty { get; }
+    public bool IsPinned { get; }
+    public int Rank { get; }
+    public ReadOnlySpan<nint> Strides { get; }
+
+    public ref T this[params ReadOnlySpan<nint> indices] { get; }
+
+    public static implicit operator TensorSpan<T>(Tensor<T> value);
+    public static implicit operator TensorReadOnlySpan<T>(Tensor<T> value);
+
+    public TensorSpan<T> AsSpan(params ReadOnlySpan<Range<nint>> ranges);
+    public ref T GetPinnableReference();
+    public Tensor<T> Slice(params ReadOnlySpan<Range<nint>> ranges);
+
+    public void Clear<T>();
+    public void CopyTo<T>(TensorSpan<T> destination);
+    public void Fill<T>(T value);
+    public bool TryCopyTo<T>(TensorSpan<T> destination);
+
+    // IEnumerable
+
+    public Enumerator GetEnumerator();
+
+    // IEquatable
+
+    public bool Equals(Tensor<T> other);
+
+    // IEqualityOperators
+
+    public static bool operator ==(Tensor<T> left, Tensor<T> right);
+    public static bool operator ==(Tensor<T> left, T right);
+
+    public static bool operator !=(Tensor<T> left, Tensor<T> right);
+    public static bool operator !=(Tensor<T> left, T right);
+
+    public struct Enumerator
+    {
+        public ref readonly T Current { get; }
+
+        public bool MoveNext();
+    }
+}
+
+public static partial class Tensor
+{
+    // Effectively mirror the TensorPrimitives surface area. Following the general pattern
+    // where we will return a new tensor and an overload that takes in the destination explicitly.
+    // * public static ITensor<T> Abs<T>(ITensor<T> x) where T : INumberBase<T>;
```

Comment:
It might be desirable to allow creating the same kind of `ITensor` and is probably worth some more discussion.

However, we aren't actually allowing the user to provide the implementation for `Abs` here (to help minimize the number of interfaces, etc) and the inputs may not all be the same (consider `MultiplyAdd` which takes in 3 different tensors). So I think such support would likely require some new kind of `ITensorAllocator` (or better name) that allows the user to customize how temporaries are created. I think that gets more into the concept of how temporaries work beyond the barebones support and is what we'll want to discuss in our offline meeting.